File: 2015/day_01/src/main.rs
use std::fs::File;
use std::io::prelude::*;

fn part_1(instructions: &str) -> isize {
    instructions.chars().fold(0, |mut floor, instruction| {
        match instruction {
            '(' => floor += 1,
            ')' => floor -= 1,
            _ => unreachable!(),
        };
        floor
    })
}

fn part_2(instructions: &str) -> isize {
    let mut pos = 0;
    let mut count = 0;
    for instr in instructions.chars() {
        count += 1;
        match instr {
            '(' => pos += 1,
            ')' => pos -= 1,
            _ => (),
        }
        if pos == -1 {
            break;
        }
    }
    count
}

fn main() -> std::io::Result<()> {
    let mut input = File::open("input.txt")?;
    let mut contents = String::new();
    input.read_to_string(&mut contents)?;

    let contents = contents.trim();

    let floor = part_1(contents);
    let position = part_2(contents);

    println!("Part 1: {floor} Part 2: {position}");

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_balanced_paren() {
        assert_eq!(0, part_1("(())"));
    }

    #[test]
    fn test_left_paren() {
        assert_eq!(3, part_1("((("));
    }

    #[test]
    fn test_right_paren() {
        assert_eq!(-3, part_1(")))"));
    }

    #[test]
    fn test_inbalanced_paren() {
        assert_eq!(3, part_1("(()(()("));
    }

    #[test]
    fn test_get_pos() {
        assert_eq!(5, part_2("()())"));
    }
}

File: 2015/day_01/main.py
def part_1(chars: str) -> int:
    floor: int = 0
    for ch in chars:
        match ch:
            case '(': floor += 1
            case ')': floor -= 1
    return floor

def part_2(chars: str) -> int:
    floor: int = 0
    for count, ch in enumerate(chars):
        if floor == -1: return count
        match ch:
            case '(': floor += 1
            case ')': floor -= 1
    return -1

def main(test = False) -> None:
    if test:
        assert part_1("(())") == 0
        assert part_1("()()") == 0
        assert part_1("(((") == 3
        assert part_1("(()(()(") == 3
        assert part_1("))(((((") == 3
        assert part_1("())") == -1
        assert part_1("))(") == -1
        assert part_1(")))") == -3
        assert part_1(")())())") == -3
        print("**part_1 tests pass**")
        
        assert part_2(")") == 1
        assert part_2("()())") == 5
        print("**part_2 tests pass**")
    else:
        with open("input.txt", "r") as f:
            content = f.read()
            print(part_1(content))
            print(part_2(content))

main(test = False)
